drivers/auxdisplay/auxdisplay_serlcd.c

```diff
@@ -0,0 +1,398 @@
+
+		return auxdisplay_serlcd_send_special_command(
+			dev, SERLCD_SPECIAL_SET_DD_RAM_ADDRESS | cursor_address);
+	}
+
+	/*
+	 * other types of movement are not implemented/supported
+	 */
+	return -ENOSYS;
+}
+
+static int auxdisplay_serlcd_cursor_position_get(const struct device *dev, int16_t *x, int16_t *y)
+{
+	const struct auxdisplay_serlcd_data *data = dev->data;
+
+	*x = (int16_t)data->cursor_x;
+	*y = (int16_t)data->cursor_y;
+
+	return 0;
+}
+
+static int auxdisplay_serlcd_capabilities_get(const struct device *dev,
+					      struct auxdisplay_capabilities *capabilities)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+
+	memcpy(capabilities, &config->capabilities, sizeof(struct auxdisplay_capabilities));
+
+	return 0;
+}
+
+static int auxdisplay_serlcd_clear(const struct device *dev)
+{
+	int rc = auxdisplay_serlcd_send_command(dev, SERLCD_COMMAND_CLEAR);
+
+	/*
+	 * clearing takes more time
+	 */
+	k_sleep(K_MSEC(10));
+	return rc;
+}
+
+static int auxdisplay_serlcd_custom_character_set(const struct device *dev,
+						  struct auxdisplay_character *character)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+
+	/*
+	 * only indexes 0..7 are supported
+	 */
+	const uint8_t char_index = character->index;
+
+	if (char_index > 7) {
+		return -EINVAL;
+	}
+
+	/*
+	 * custom characters are accessible via char codes 0x08..0x0f
+	 */
+	character->character_code = 0x08 | char_index;
+
+	int rc = auxdisplay_serlcd_send_command(dev, SERLCD_COMMAND_SET_CUSTOM_CHAR + char_index);
+
+	if (!rc) {
+		return rc;
+	}
+
+	/*
+	 * the display expects the custom character as 8 lines of 5 bit each, shades are not
+	 * supported
+	 */
+	for (int l = 0; l < 8; ++l) {
+		uint8_t buffer = 0;
+
+		for (int i = 0; i < 5; ++i) {
+			if (character->data[(l * 5) + i]) {
+				buffer |= BIT(4 - i);
+			}
+		}
+		rc = i2c_write_dt(&config->bus, &buffer, 1);
+		if (!rc) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+static void auxdisplay_serlcd_advance_current_position(const struct device *dev)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	struct auxdisplay_serlcd_data *data = dev->data;
+
+	++(data->cursor_x);
+	if (data->cursor_x >= config->capabilities.columns) {
+		data->cursor_x = 0;
+		++(data->cursor_y);
+	}
+	if (data->cursor_y >= config->capabilities.rows) {
+		data->cursor_y = 0;
+	}
+}
+
+static int auxdisplay_serlcd_write(const struct device *dev, const uint8_t *text, uint16_t len)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+
+	int rc = 0;
+
+	/*
+	 * the display wraps around by itself, just write the text and update the position data
+	 */
+	for (int i = 0; i < len; ++i) {
+		uint8_t character = text[i];
+
+		/*
+		 * customer characters require a special command, so check for custom char
+		 */
+		if ((character & 0xf8) == 0x08) {
+			rc = auxdisplay_serlcd_send_command(dev, SERLCD_COMMAND_WRITE_CUSTOM_CHAR +
+									 (character & 0x07));
```

Comment:
drop the blank line

Reply:
Which blank line?

Reply:
Oh, sorry misread the diff. You could use an intermediate variable for `SERLCD_COMMAND_WRITE_CUSTOM_CHAR + (character & 0x07)`? Would save this line from wrapping, the indentation is also a bit odd.

Reply:
Also `~SERLCD_CUSTOM_CHAR_BITMASK` instead of 0x07?

Reply:
Done